ros2: Porting the code to ros2

Fildo7525 opened the pull request:
This is an incomplete solution. Just for now.

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
ros2: Porting the code to ros2

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

MichalDobis commented on src/blackmetal/CMakeLists.txt:
```diff
@@ -0,0 +1,86 @@
+	blackmetal
+PUBLIC
+	c_std_99
+	cxx_std_17
```
toto asi nepotrebujes ked setujes 17 v set(CMAKE_CXX_STANDARD 17)

Fildo7525 replied:
To su defaultne konfiguracie co vyhodi ```ros2 pkg create <package_name> --node-name <node_name>```

Fildo7525 replied:
Mozem vvyskusat ci to pojde aj bez toho popripade to potom hodim naspat

Fildo7525 replied:
eb022081df89aecde376f6230586632b2d02ae3e slo to tak to necham tak ako si navrhol

MichalDobis commented on src/blackmetal/src/communication/BlackMetal.hpp:
```diff
@@ -0,0 +1,47 @@
+ * The class creates client and returns this instance on every call.
+ */
+class BlackMetal
+	: public rclcpp::Node
+	, public Client
```
nebolo by vhodnejsie tieto objekty mat ako private members?

MichalDobis replied:
Ci to je povinnost pri ros2?

Fildo7525 replied:
Co som pozeral viacero probramov tak vsade sa inheritencia riesila cez public. Pozrel som si aj tak Nodu ze co by to ovplyvnilo a je tam jeden konstruktor, ktory by sa stal privatnym.

MichalDobis commented on src/blackmetal/src/communication/Client.cpp:
```diff
@@ -0,0 +1,123 @@
+	return std::get<bm::Status>(buffer);
+}
+
+std::string Client::recieve()
```
```suggestion
std::string Client::receive()
```

Fildo7525 replied:
chyba v spellingu opravil som to uz

MichalDobis commented on src/blackmetal/src/communication/Client.cpp:
```diff
@@ -0,0 +1,123 @@
+	return msg.substr(begin+1, end-1);
+}
+
+bm::Status Client::evalReturnState(const std::string &returnJson)
+{
+	std::string tmp = retrieveAnswer(returnJson);
+	INFO(tmp);
+
+	if (tmp == "RECIEVE_OK") {
+		INFO("The execution ran correctly");
+		return bm::Status::OK;
+	} else {
+		ERR("The rebot buffer is full. The send data will not be used.");
+		return bm::Status::FULL_BUFFER;
+	}
+}
```
Tuto funkcionalitu by client nemal robit. Client by mal iba citat a zapisovat data na socket. Vyssia vrstva, ktora pouziva clienta by mala parsovat status

Fildo7525 replied:
Vyriesil som to tak ze v klientovi som nechal pure virtualnu funkciu a v childe ju prepisujem.

MichalDobis commented on src/blackmetal/src/communication/Client.hpp:
```diff
@@ -0,0 +1,77 @@
+ * The robot receives the json string and returns a json string with a specific return status.
+ * This status is than mapped on BlackMetal::Status enum type.
+ */
+enum class Status {
```
Header by som rozdelil na Client, ktory iba implementuje komunikaciu a druhy header, ktory parsuje

Fildo7525 replied:
Rozdelil som to na dva subory ale nevidim zmysel to davat do ineho suboru kedze tento client bude pracovat len s tymito enumami a nicim inym. Keby ze to malo byt genericke tak by som to spravil inak. Ako to vidis ty?

MichalDobis replied:
Osobne mi to nevadi, len z mojej skusenosti odporucam vzdy sa drzat "single-responsibility principle", cize jeden objekt by mal robit len jednu vec. Preto by som to rozdeloval na komunikaciu a na parsovanie.

Commit:
communication: Extract server

Extract server part of the BlackMetal class and create
a client base class for creating communication link with the robot.

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
logging: Multiple modules logging

The error of using multiple instances of the same variable vas solved
with static keyword.

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
cp: spelling mistake

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
enums: Macro definition in separate file

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
csw: Response evaluation made virtual

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
comunication: private Client inheritance

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
csw: Removed unused destrucor

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
build: Run compiled nodes

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
launch: the execution of the nodes made easier

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
launch: Program run method changed

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: Test logging removed

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
cmake: Include path extedned

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
cmake: remove doubled compile standart setup

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: safer way to log messages

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: Make logging available only in debug mode

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: Using compile time to differ logs

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
csw: Received message parsing

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
csw: Received message parsing

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: Adding comments

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: including log level to the message

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: Separate send and receive methods

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
lsw: receive messages via timer

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
cache: Ignoring pycache from launch file

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
launch: log: Realtime log printing

We have to use flush on every log because of launch file.
There is an inner buffer that chatches the output. If the launch file
was not used there would be no need for flush at every write.

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
csw: Calculation the right and left wheel velocities

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
comments: Creating comments for doxygen documentation

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
doc: Doxygen config generation

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
cp: spell: recieve -> receive

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
ignore: Ignoring generated documents

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
update

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Fildo7525 committed:
Merge pull request #3 from Fildo7525/logging

Logging

Fildo7525 committed:
Merge branch 'port-to-ros2' into launch

Fildo7525 committed:
Merge pull request #2 from Fildo7525/launch

Launch

Fildo7525 committed:
Update BlackMetal.cpp

Fildo7525 committed:
Delete BlackMetal.cpp
